src/libsync/discovery.cpp

```diff
@@ -230,25 +182,41 @@ void ProcessDirectoryJob::process()
 #endif
         if (handleExcluded(path._target, e.localEntry.name,
                 e.localEntry.isDirectory || e.serverEntry.isDirectory, isHidden,
-                e.localEntry.isSymLink || isServerEntryWindowsShortcut))
+                e.localEntry.isSymLink || isServerEntryWindowsShortcut, e.serverEntry.isValid() || e.dbEntry.isValid()))
             continue;
 
         if (_queryServer == InBlackList || _discoveryData->isInSelectiveSyncBlackList(path._original)) {
             processBlacklisted(path, e.localEntry, e.dbEntry);
             continue;
         }
-        if (!checkForInvalidFileName(path, entries, e)) {
-            continue;
-        }
         processFile(std::move(path), e.localEntry, e.serverEntry, e.dbEntry);
     }
     QTimer::singleShot(0, _discoveryData, &DiscoveryPhase::scheduleMoreJobs);
 }
 
-bool ProcessDirectoryJob::handleExcluded(const QString &path, const QString &localName, bool isDirectory, bool isHidden, bool isSymlink)
+bool ProcessDirectoryJob::handleExcluded(const QString &path, const QString &localName, bool isDirectory, bool isHidden, bool isSymlink, bool isNotJustCreated)
 {
     auto excluded = _discoveryData->_excludes->traversalPatternMatch(path, isDirectory ? ItemTypeDirectory : ItemTypeFile);
 
+    const auto fileName = path.mid(path.lastIndexOf('/') + 1);
+
+    if (excluded == CSYNC_NOT_EXCLUDED) {
+        const auto endsWithSpace = fileName.endsWith(QLatin1Char(' '));
+        const auto startsWithSpace = fileName.startsWith(QLatin1Char(' '));
+        if (startsWithSpace && endsWithSpace) {
+            excluded = CSYNC_FILE_EXCLUDE_LEADING_AND_TRAILING_SPACE;
+        } else if (endsWithSpace) {
+            excluded = CSYNC_FILE_EXCLUDE_TRAILING_SPACE;
+        } else if (startsWithSpace) {
+            excluded = CSYNC_FILE_EXCLUDE_LEADING_SPACE;
+        }
+    }
+
+    if ((excluded == CSYNC_FILE_EXCLUDE_LEADING_SPACE || excluded == CSYNC_FILE_EXCLUDE_TRAILING_SPACE || excluded == CSYNC_FILE_EXCLUDE_LEADING_AND_TRAILING_SPACE)
+            && ((!Utility::isWindows() && isNotJustCreated) || _discoveryData->_leadingAndTrailingSpacesFilesAllowed.contains(_discoveryData->_localDir + path))) {
```

Comment:
@mgallien

- First of all, on Windows - we never allow to sync files that contain leading/trailing spaces, so, we do not remove the excluded value (`CSYNC_FILE_EXCLUDE_LEADING_SPACE`, `CSYNC_FILE_EXCLUDE_TRAILING_SPACE`, `CSYNC_FILE_EXCLUDE_LEADING_AND_TRAILING_SPACE`). The warning must always get displayed on Windows because in this design, we do not allow accepting invalid file names (with leading or trailing spaces).
- On other OSs (Linux, macOS), we can safely download the leading/trailing spaces file from the server and we will be allowed to do it via the popup, such that these file path gets inserted into `_discoveryData->_leadingAndTrailingSpacesFilesAllowed`, meaning that, during next sync, we will allow those to get downloaded. The same happens with file that we have just created locally on Linux/macOS - we can allow it to get uploaded without removing the spaces, using the same `_discoveryData->_leadingAndTrailingSpacesFilesAllowed` list.

P.S. I have changed the name of `isNotJustCreated` to be, hopefully, more understandable.